### PR TITLE
many: arbitrary homedir support via snap configuration

### PR DESCRIPTION
**Pull request by mardy**
This adds support for arbitrary homedir paths via the new `snap set system homedirs=/remote/users` option.

Uses https://github.com/snapcore/snapd/pull/11367

### CODE REVIEW

File: cmd/snap-confine/ns-support.c [hunk 1]
```diff
@@ -319,6 +346,30 @@ static bool should_discard_current_ns(dev_t base_snap_dev)
...
+
+	mi = sc_parse_mountinfo(NULL);
+	if (mi == NULL) {
+		die("cannot parse mountinfo of the current process");
+	}
+
+	// The namespace may become "stale" when the rootfs is not the same
+	// device we found above. This will happen whenever the base snap is
+	// refreshed since the namespace was first created.
+	if (base_snap_device_changed(mi, base_snap_dev)) return true;
```
Comment (mvo5) [inline cmd/snap-confine/ns-support.c:363]: (nitpick) maybe our code formatter did that, but I would prefer if this return would be in a new line (and below)

File: interfaces/apparmor/backend.go [hunk 2]
```diff
@@ -101,72 +101,15 @@ func (b *Backend) Initialize(opts *interfaces.SecurityBackendOptions) error {
...
-
-	// Ensure that generated policy is what we computed above.
-	created, removed, err := osutil.EnsureDirState(dirs.SnapConfineAppArmorDir, glob, policy)
-	if err != nil {
-		return fmt.Errorf("cannot synchronize snap-confine policy: %s", err)
-	}
-	if len(created) == 0 && len(removed) == 0 {
-		// If the generated policy didn't change, we're all done.
-		return nil
+	if _, err := apparmor_sandbox.SetupSnapConfineSnippets([]string{}); err != nil {
```
Comment (mvo5) [inline interfaces/apparmor/backend.go:111]: It's probably more idiomatic to use `nil` here instead of `[]string{}`?

**Review by mvo5 — COMMENTED** [review]
Thanks, this looks very good, two nitpicks so far but I need to give it another look

File: interfaces/apparmor/backend.go (hunk 2 above)
Comment (Meulengracht) [reply interfaces/apparmor/backend.go]: Good point, I've updated this to pass nil instead. `SetupSnapConfineSnippets` seems to handle nil just fine.

File: cmd/libsnap-confine-private/cleanup-funcs-test.c [hunk 3]
```diff
@@ -142,6 +143,31 @@ static void test_cleanup_close(void)
...
+{
+	/* It is safe to use with a NULL pointer */
+	sc_cleanup_deep_strv(NULL);
+
+	char **argses = NULL;
+	/* It is OK if the pointer value is NULL */
+	sc_cleanup_deep_strv(&argses);
+	g_assert_null(argses);
+
+	/* It is OK if the string pointers are NULL */
```
Comment (pedronis) [inline cmd/libsnap-confine-private/cleanup-funcs-test.c:156]: the comment is a bit unclear as NULL is the terminator for the vector if I understand the code

File: sandbox/apparmor/profile.go [hunk 4]
```diff
@@ -251,3 +302,101 @@ var SnapConfineDistroProfilePath = func() string {
 
 	return ""
 }
+
+// SetupSnapConfineSnippets inspects the system and sets up local apparmor
+// policy for snap-confine.
+// Local policy is included by the system-wide policy. If the local policy
+// has changed then the apparmor profile for snap-confine is reloaded.
```
Comment (pedronis) [inline sandbox/apparmor/profile.go:309]: I don't understand the last bit of the comment, as I don't see reloading code in this function. It seems reloading is a separate function?

File: sandbox/apparmor/profile.go [hunk 5]
```diff
@@ -51,7 +53,40 @@ const (
...
 )
 
-var runtimeNumCPU = runtime.NumCPU
+var (
+	runtimeNumCPU = runtime.NumCPU
+
+	osutilIsHomeUsingNFS        = osutil.IsHomeUsingNFS
+	osutilIsRootWritableOverlay = osutil.IsRootWritableOverlay
+)
+
```
Comment (pedronis) [inline sandbox/apparmor/profile.go:62]: most of the new code in this file doesn't seem to have matching unit tests in the package? do we need some tests copied over from the backend?

File: sandbox/apparmor/profile.go [hunk 6]
```diff
@@ -251,3 +302,101 @@ var SnapConfineDistroProfilePath = func() string {
...
+}
+
+// Reload the profile of snap-confine. This function must be called when the
+// AppArmor profile snippets for snap-confine have been changed. Note that
+// there might be more than one of them in the system, because besides the one
+// coming with the system installation, also the snapd and core snaps ship a
+// snap-confine binary. Of course, only one of these binaries will be used by
+// snapd, but in this context we don't want to deal with this logic, so let's
+// just reload all of them.
+var ReloadSnapConfineProfile = func() error {
```
Comment (pedronis) [inline sandbox/apparmor/profile.go:386]: I'm confused, I don't see calls to this?

File: interfaces/apparmor/backend.go [hunk 7]
```diff
@@ -101,72 +101,15 @@ func (b *Backend) Initialize(opts *interfaces.SecurityBackendOptions) error {
...
-	created, removed, err := osutil.EnsureDirState(dirs.SnapConfineAppArmorDir, glob, policy)
-	if err != nil {
-		return fmt.Errorf("cannot synchronize snap-confine policy: %s", err)
-	}
-	if len(created) == 0 && len(removed) == 0 {
-		// If the generated policy didn't change, we're all done.
-		return nil
+	if _, err := apparmor_sandbox.SetupSnapConfineSnippets(nil); err != nil {
+		return err
 	}
```
Comment (pedronis) [inline interfaces/apparmor/backend.go:113]: this has still its own code to reload the profile if I see things correctly?

File: cmd/snap-confine/snap-confine-invocation.c [hunk 8]
```diff
@@ -144,3 +145,61 @@ void sc_check_rootfs_dir(sc_invocation *inv) {
...
+        if (homedir[0] == '\0') {
+            /* The typical case where we'd get an empty string is
+             * if the config file just contains
+             *
+             * homedirs=
+             */
+            continue;
+        }
+        inv->homedirs[current_index++] = sc_strdup(homedir);
+        homedir = strtok_r(NULL, "/", &buf_saveptr);
```
Comment (pedronis) [inline cmd/snap-confine/snap-confine-invocation.c:203]: I'm not sure why the delimiter is "/" here?

File: cmd/snap-confine/mount-support.c [hunk 9]
```diff
@@ -203,12 +203,34 @@ struct sc_mount_config {
...
 	const char *base_snap_name;
 	const char *snap_instance;
 };
 
+static void sc_create_mount_points(const char *scratch_dir,
+                                   const struct sc_mount *mounts)
+{
+	char dst[PATH_MAX] = { 0 };
+	sc_identity old = sc_set_effective_identity(sc_root_group_identity());
+	// Bind mount certain directories from the host filesystem to the scratch
```
Comment (pedronis) [inline cmd/snap-confine/mount-support.c:219]: this comment is not super clear, it sounds like the code here does the bind mounts but actually is just creating the target for those. Maybe we need to reformulate.

File: cmd/snap-confine/mount-support.c [hunk 10]
```diff
@@ -919,12 +976,15 @@ void sc_populate_mount_ns(struct sc_apparmor *apparmor, int snap_update_ns_fd,
 		struct sc_mount_config normal_config = {
 			.rootfs_dir = inv->rootfs_dir,
 			.mounts = mounts,
+			.dynamic_mounts = sc_create_homedir_mounts(inv),
```
Comment (pedronis) [inline cmd/snap-confine/mount-support.c:979]: this seems to merit a comment

File: cmd/snap-confine/mount-support.c [hunk 11]
```diff
@@ -534,6 +556,8 @@ static void sc_bootstrap_mount_namespace(const struct sc_mount_config *config)
 	// below serves as the foundation of the mount sandbox.
 	sc_do_mount("none", scratch_dir, NULL, MS_REC | MS_SLAVE, NULL);
 	sc_do_mounts(scratch_dir, config->mounts);
+	sc_create_mount_points(scratch_dir, config->dynamic_mounts);
+	sc_do_mounts(scratch_dir, config->dynamic_mounts);
```
Comment (pedronis) [inline cmd/snap-confine/mount-support.c:560]: also this code

File: cmd/snap-confine/ns-support.c [hunk 12]
```diff
@@ -273,20 +273,47 @@ static dev_t find_base_snap_device(const char *base_snap_name,
 	return base_snap_dev;
 }
 
-static bool should_discard_current_ns(dev_t base_snap_dev)
+static bool homedirs_are_mounted(sc_mountinfo *mi, char **homedirs)
 {
```
Comment (pedronis) [inline cmd/snap-confine/ns-support.c:277]: the diff of this file is quite messy, maybe there's a reording of the changes/additions to avoid this?

**Review by pedronis — COMMENTED** [review]
did a pass, the changes in sandbox/apparmor are a bit confusing as there's untested/unused code, unclear if we need more refactoring elsewhere or if some of the new code is not needed

File: cmd/libsnap-confine-private/cleanup-funcs-test.c (hunk 3 above)
Comment (Meulengracht) [reply cmd/libsnap-confine-private/cleanup-funcs-test.c:156]: Good point, I'll clear this comment up

File: sandbox/apparmor/profile.go (hunk 4 above)
Comment (Meulengracht) [reply sandbox/apparmor/profile.go:309]: I'll remove that part of the comment, also there is currently no need for the function to be a `var` as it's not mocked currently.

File: sandbox/apparmor/profile.go (hunk 5 above)
Comment (Meulengracht) [reply sandbox/apparmor/profile.go:62]: Nice catch, I'll look into this

File: sandbox/apparmor/profile.go (hunk 6 above)
Comment (Meulengracht) [reply sandbox/apparmor/profile.go:386]: It seems to be a unfinished part of refactoring some code from backend to profile. I'll remove this right now and do this in a followup, I don't believe there is any need for this to be a part of this PR

File: cmd/snap-confine/snap-confine-invocation.c (hunk 8 above)
Comment (Meulengracht) [reply cmd/snap-confine/snap-confine-invocation.c:203]: It's a bug, should have been `,`

File: cmd/snap-confine/mount-support.c (hunk 9 above)
Comment (Meulengracht) [reply cmd/snap-confine/mount-support.c:219]: Yea you are right, the comment does not match the code. I'll rewrite the docs for this func

File: cmd/snap-confine/ns-support.c (hunk 12 above)
Comment (Meulengracht) [reply cmd/snap-confine/ns-support.c]: Managed to clear this diff up

File: snap/sysparams/sysparams.go [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+	params, err := parseSystemParams(string(data))
+	if err != nil {
+		return params, err
+	}
+	params.path = path
+	return params, nil
+}
+
+// Update flushes the changes (if any) to path supplied to Open.
+func (ssp *SnapdSystemParams) Update() error {
```
Comment (pedronis) [inline snap/sysparams/sysparams.go:87]: I think this would be more typically called Write (seed Modeenv for example)

File: snap/sysparams/sysparams.go [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+	contents := fmt.Sprintf("homedirs=%s\n", ssp.Homedirs)
+	desiredState := &osutil.MemoryFileState{
+		Content: []byte(contents),
+		Mode:    0644,
+	}
+	if err := osutilEnsureFileState(ssp.path, desiredState); errors.Is(err, osutil.ErrSameState) {
```
Comment (pedronis) [inline snap/sysparams/sysparams.go:93]: given that this is tiny and we don't return anything about whether is was updated or not I wonder if it shouldn't simply use  osutil.AtomicWriteFile?

File: cmd/snap-confine/snap-confine-invocation.c [hunk 15]
```diff
@@ -144,3 +145,66 @@ void sc_check_rootfs_dir(sc_invocation *inv) {
...
+        if (homedir[0] == '\0') {
+            /* The typical case where we'd get an empty string is
+             * if the config file just contains
+             *
+             * homedirs=
+             */
+            continue;
+        }
+        inv->homedirs[current_index++] = sc_strdup(homedir);
+        homedir = strtok_r(NULL, ",", &buf_saveptr);
```
Comment (pedronis) [inline cmd/snap-confine/snap-confine-invocation.c:208]: should this also use "\n" in delim?

File: cmd/snap-confine/mount-support.c [hunk 16]
```diff
@@ -878,6 +903,41 @@ static bool __attribute__((used))
 	return false;
 }
 
+static struct sc_mount *sc_create_homedir_mounts(const struct sc_invocation *inv)
```
Comment (pedronis) [inline cmd/snap-confine/mount-support.c:906]: it seems create is used mostly for things happening to the system but this is only setting things, so mabye s/create/init/ ?

File: sandbox/apparmor/profile.go [hunk 17]
```diff
@@ -251,3 +303,86 @@ var SnapConfineDistroProfilePath = func() string {
 
 	return ""
 }
+
+var UserHomedirs = func() ([]string, error) {
```
Comment (pedronis) [inline sandbox/apparmor/profile.go:307]: I don't think this should be exported, also not sure why the User part in the name?

File: sandbox/apparmor/profile.go [hunk 18]
```diff
@@ -251,3 +303,86 @@ var SnapConfineDistroProfilePath = func() string {
...
+	if err != nil {
+		return nil, err
+	}
+	return strutil.CommaSeparatedList(ssp.Homedirs), nil
+}
+
+// SetupSnapConfineSnippets inspects the system and sets up local apparmor
+// policy for snap-confine. Local policy is included by the system-wide policy.
+// Returns whether any modifications was made to the snap-confine snippets.
+func SetupSnapConfineSnippets() (wasChanged bool, err error) {
```
Comment (pedronis) [inline sandbox/apparmor/profile.go:319]: moving these here should probably be done separately as this PR is rather largish and mixing refactorings and new code is a bit messy

File: sandbox/apparmor/profile.go [hunk 19]
```diff
@@ -251,3 +303,86 @@ var SnapConfineDistroProfilePath = func() string {
...
+
+	// Ensure that generated policy is what we computed above.
+	created, removed, err := osutil.EnsureDirState(dirs.SnapConfineAppArmorDir, glob, policy)
+	if err != nil {
+		return false, fmt.Errorf("cannot synchronize snap-confine policy: %s", err)
+	}
+	return len(created) > 0 || len(removed) > 0, nil
+}
+
+func RemoveSnapConfineSnippets() error {
```
Comment (pedronis) [inline sandbox/apparmor/profile.go:384]: this doesn't seem to be tested in the package

File: overlord/configstate/configcore/homedirs.go [hunk 20]
```diff
@@ -104,21 +102,17 @@ func updateHomedirsConfig(config string, opts *fsOnlyContext) error {
...
+	if err != nil {
+		return err
 	}
-	if err := osutilEnsureFileState(configPath, desiredState); errors.Is(err, osutil.ErrSameState) {
-		// The state is unchanged, nothing to do
-		return nil
-	} else if err != nil {
+	ssp.Homedirs = config
+	if err := ssp.Update(); err != nil {
 		return err
```
Comment (pedronis) [inline overlord/configstate/configcore/homedirs.go:116]: this change and introducing sysparams could also be its own PR

**Review by pedronis — COMMENTED** [review]
did a pass, I think a couple of things would be better split out from this making for a smaller and more readable final PR

File: cmd/snap-confine/snap-confine-invocation.c (hunk 15 above)
Comment (Meulengracht) [reply cmd/snap-confine/snap-confine-invocation.c]: Yes! Very nice catch

File: cmd/snap-confine/mount-support.c (hunk 16 above)
Comment (Meulengracht) [reply cmd/snap-confine/mount-support.c]: Renamed this, thank you for the review

File: sandbox/apparmor/profile.go (hunk 17 above)
Comment (Meulengracht) [reply sandbox/apparmor/profile.go]: Ah right, that's correct

File: sandbox/apparmor/profile.go [hunk 21]
```diff
@@ -287,6 +304,14 @@ var SnapConfineDistroProfilePath = func() string {
 	return ""
 }
 
+var loadHomedirs = func() ([]string, error) {
+	ssp, err := sysparams.Open("")
+	if err != nil {
+		return nil, err
+	}
+	return strutil.CommaSeparatedList(ssp.Homedirs), nil
+}
```
Comment (pedronis) [inline sandbox/apparmor/profile.go:313]: this functions ends up always mocked so it is never exercised, it needs a dedicated test

File: overlord/configstate/configcore/homedirs.go [hunk 22]
```diff
@@ -88,6 +82,10 @@ func configureHomedirsInAppArmorAndReload(homedirs []string, opts *fsOnlyContext
 
 	// Only reload profiles if it's during runtime
 	if opts == nil {
+		if _, err := apparmorSetupSnapConfineSnippets(); err != nil {
+			return err
+		}
```
Comment (pedronis) [inline overlord/configstate/configcore/homedirs.go:87]: the comment before the if is not quite right now, because this bit is writing an apparmor snippet and not purely about reloading, I think it's ok not to write it early because the snippet is only useful once snaps are installed and not earlier than that but it needs to be clarified. We probably need a spread test that sets this via gadget defaults and sees that things work as expected for snaps installed during first boot

File: overlord/configstate/configcore/homedirs_test.go [hunk 23]
```diff
@@ -264,6 +220,11 @@ func (s *homedirsSuite) TestConfigureHomedirsHappy(c *C) {
 	})
 	defer restore()
 
+	restore = configcore.MockApparmorSetupSnapConfineSnippets(func() (bool, error) {
+		return false, nil
```
Comment (pedronis) [inline overlord/configstate/configcore/homedirs_test.go:224]: don't we also need to check that this gets called as needed?

File: cmd/snap-confine/mount-support.c [hunk 24]
```diff
@@ -878,6 +903,41 @@ static bool __attribute__((used))
 	return false;
 }
 
+static struct sc_mount *sc_init_homedir_mounts(const struct sc_invocation *inv)
```
Comment (pedronis) [inline cmd/snap-confine/mount-support.c:906]: sorry, I somehow thought this was setting things but it actually returns them, so it can be called just sc_homedir_mounts

File: overlord/configstate/configcore/homedirs.go (hunk 22 above)
Comment (pedronis) [reply overlord/configstate/configcore/homedirs.go:87]: otoh might that mostly not relevant because this option should work only classic, we don't have a concept of classic only options but maybe this should be one? cc @mvo5

**Review by pedronis — COMMENTED** [review]
did a full pass, some wondering about the option behavior if used through defaults but also whether that make sense at all

File: overlord/configstate/configcore/homedirs_test.go (hunk 23 above)
Comment (Meulengracht) [reply overlord/configstate/configcore/homedirs_test.go:226]: Good point, thanks!

File: cmd/snap-confine/mount-support.c (hunk 24 above)
Comment (Meulengracht) [reply cmd/snap-confine/mount-support.c]: That makes sense, thank you for catching that

File: sandbox/apparmor/profile.go (hunk 21 above)
Comment (Meulengracht) [reply sandbox/apparmor/profile.go:313]: Really good point about this code not being hit! I don't think function deserves it's own dedicated test due to the nature of the function. Instead what I've done is make sure this code is excercised by the happy case `TestSetupSnapConfineSnippetsHomedirs` where it loads the actual system-params instead of mocking this function in all the test cases.

File: cmd/snap-confine/mount-support.c [hunk 25]
```diff
@@ -878,6 +903,41 @@ static bool __attribute__((used))
...
+	struct sc_mount *current_mount = mounts;
+	for (char **path = inv->homedirs; *path != NULL; path++) {
+		debug("Adding homedir: %s", *path);
+		current_mount->path = *path;
+		current_mount++;
+	}
+	return mounts;
+}
+
+static void sc_free_dynamic_mounts(struct sc_mount *mounts)
```
Comment (mvo5) [inline cmd/snap-confine/mount-support.c:934]: (super nitpick) I wonder if we should express in the function name that it does not do a "deep" free, maybe `sc_free_dynamic_mounts_shallow()` or `sc_free_dynamic_mounts_struct_only()` ? [and yay, rust to the rescue here]

File: cmd/snap-confine/snap-confine-invocation.c [hunk 26]
```diff
@@ -144,3 +145,66 @@ void sc_check_rootfs_dir(sc_invocation *inv) {
 
     die("cannot locate base snap %s", inv->base_snap_name);
 }
+
+static char* read_homedirs_from_system_params(void)
+{
+    FILE *f SC_CLEANUP(sc_cleanup_file) = NULL;
+    f = fopen("var/lib/snapd/system-params", "r");
```
Comment (mvo5) [inline cmd/snap-confine/snap-confine-invocation.c:152]: Is it deliberate that this is a relative path? If so maybe a comment about the rational?

File: cmd/snap-confine/mount-support.c [hunk 27]
```diff
@@ -878,6 +903,41 @@ static bool __attribute__((used))
...
+	// We add one element for the end-of-array indicator.
+	struct sc_mount *mounts = calloc(num_homedirs + 1, sizeof(struct sc_mount));
+	if (mounts == NULL) {
+		die("cannot allocate mount data for homedirs");
+	}
+
+	struct sc_mount *current_mount = mounts;
+	for (char **path = inv->homedirs; *path != NULL; path++) {
+		debug("Adding homedir: %s", *path);
+		current_mount->path = *path;
```
Comment (mvo5) [inline cmd/snap-confine/mount-support.c:928]: (super-duper-nitpick) I was wondering if we should stdup() here just to make reasoning who owns the memory easier but all of this is very local so I'm probably overthiking things [and again - rust to the rescue :)]

File: cmd/snap-confine/snap-confine-invocation.c [hunk 28]
```diff
@@ -144,3 +145,66 @@ void sc_check_rootfs_dir(sc_invocation *inv) {
...
+     * right number of strings, let's count how many commas we have.
+     */
+    int num_commas = 0;
+    for (char *c = config_line; *c != '\0'; c++) {
+        if (*c == ',') {
+            num_commas++;
+        }
+    }
+
+    /* We add *two* elements here: one is because of course the number of
```
Comment (mvo5) [inline cmd/snap-confine/snap-confine-invocation.c:184]: Love this comment

File: cmd/snap-confine/snap-confine-invocation.c [hunk 29]
```diff
@@ -144,3 +145,66 @@ void sc_check_rootfs_dir(sc_invocation *inv) {
...
+
+    // strtok_r needs a pointer to keep track of where it is in the
+    // string.
+    char *buf_saveptr = NULL;
+
+    int current_index = 0;
+    char *homedir = strtok_r(config_line, ",\n", &buf_saveptr);
+    while (homedir != NULL) {
+        if (homedir[0] == '\0') {
+            /* The typical case where we'd get an empty string is
```
Comment (mvo5) [inline cmd/snap-confine/snap-confine-invocation.c:200]: (super nitpick) I would make this comment shorter, something like: `// Deal with cases like "homedirs=", "homedirs=,," etc` (it's nice but takes too much room IMHO for what it does explains :)

**Review by mvo5 — APPROVED** [review]
This looks super nice and is also quite reviewable now that all the prerequisite got merged. I have a bunch of nitpicks but nothing is a blocker and probably mostly style/preferences - hope it's still useful.

File: overlord/configstate/configcore/homedirs.go (hunk 22 above)
Comment (Meulengracht) [reply overlord/configstate/configcore/homedirs.go:87]: Are we refering just to `apparmorSetupSnapConfineSnippets` that we want to do only on classic?

File: cmd/snap-confine/mount-support.c [hunk 30]
```diff
@@ -878,6 +903,41 @@ static bool __attribute__((used))
 	return false;
 }
 
+static struct sc_mount *sc_homedir_mounts(const struct sc_invocation *inv)
+{
+	if (inv->homedirs == NULL) {
+		// Return empty array, but never NULL as functions rely
+		// on the mounts array to be non-NULL
+		return calloc(1, sizeof(struct sc_mount));
```
[the remark this answers is not in the document]
Comment (Meulengracht) [reply cmd/snap-confine/mount-support.c]: Ah really good idea! Added this

File: cmd/snap-confine/mount-support.c (hunk 25 above)
Comment (Meulengracht) [reply cmd/snap-confine/mount-support.c:937]: We now strdup and free them (per your other suggestion), so keeping name as-is!

File: cmd/snap-confine/mount-support.c (hunk 27 above)
Comment (Meulengracht) [reply cmd/snap-confine/mount-support.c]: I think this makes much better sense, the thought had struct me when I went through this code, never did it, so really happy you suggested this!

File: cmd/snap-confine/snap-confine-invocation.c (hunk 29 above)
Comment (Meulengracht) [reply cmd/snap-confine/snap-confine-invocation.c]: Absolutely agree! It's more sensible to be shorter about this

File: sandbox/apparmor/profile.go [hunk 31]
```diff
@@ -336,6 +361,15 @@ func SetupSnapConfineSnippets() (wasChanged bool, err error) {
 		}
 	}
 
+	if homedirs, err := loadHomedirs(); err != nil {
+		logger.Noticef("cannot determine if any homedirs are set: %v", err)
```
[the remark this answers is not in the document]
Comment (Meulengracht) [reply sandbox/apparmor/profile.go:365]: I don't know why we do this on a best-effort basis, but it's something we could bring up if this should change?

File: overlord/configstate/configcore/homedirs.go (hunk 22 above)
Comment (pedronis) [reply overlord/configstate/configcore/homedirs.go:87]: no, we mean to ignore (what we do for core only options on classic, for this we would need a new classicOnly flag) or error for the entire option on core.

Comment (Meulengracht) [reply overlord/configstate/configcore/homedirs.go:87]: Is this somewhat what you had in mind? cd3f3b7

File: cmd/snap-confine/mount-support.c [hunk 32]
```diff
@@ -878,6 +903,48 @@ static bool __attribute__((used))
...
+	struct sc_mount *mounts = calloc(num_homedirs + 1, sizeof(struct sc_mount));
+	if (mounts == NULL) {
+		die("cannot allocate mount data for homedirs");
+	}
+
+    // Copy inv->homedirs to the mount structures
+    for (int i = 0; i < num_homedirs; i++) {
+        debug("Adding homedir: %s", inv->homedirs[i]);
+        mounts[i].path = sc_strdup(inv->homedirs[i]);
+    }
```
Comment (pedronis) [inline cmd/snap-confine/mount-support.c:933]: the indentation of this looks off

File: overlord/configstate/configcore/homedirs.go [hunk 33]
```diff
@@ -88,6 +82,10 @@ func configureHomedirsInAppArmorAndReload(homedirs []string, opts *fsOnlyContext
 
 	// Only reload profiles if it's during runtime
```
Comment (pedronis) [inline overlord/configstate/configcore/homedirs.go:83]: this needs updating, // Only update apparmor configuration snippet and reload profiles if ...

**Review by pedronis — APPROVED** [review]
thanks, couple small comments

File: cmd/snap-confine/mount-support.c (hunk 32 above)
Comment (Meulengracht) [reply cmd/snap-confine/mount-support.c]: Thank you for catching that!

File: overlord/configstate/configcore/homedirs.go (hunk 33 above)
Comment (Meulengracht) [reply overlord/configstate/configcore/homedirs.go]: Updated!

**Comment by mvo5** [comment]
Fwiw, the error on core-16 looks real, looks like snap-confine segfaults during the "pc" configure hook

**Comment by mvo5** [comment]
> Fwiw, the error on core-16 looks real, looks like snap-confine segfaults during the "pc" configure hook

I found and fixed the bug, legacy mode in uc16 was not setting `sc_mount_config.dynamic_mounts` and that caused snap-confine to crash. This raises the question if the code insc_create_mount_points() and sc_do_mounts() should also be robust against NULL ?

**Comment by Meulengracht** [comment]
> > Fwiw, the error on core-16 looks real, looks like snap-confine segfaults during the "pc" configure hook
> 
> I found and fixed the bug, legacy mode in uc16 was not setting `sc_mount_config.dynamic_mounts` and that caused snap-confine to crash. This raises the question if the code insc_create_mount_points() and sc_do_mounts() should also be robust against NULL ?

I think it makes more sense to be robust against NULL in this case.

File: cmd/snap-confine/mount-support.c [hunk 34]
```diff
@@ -878,6 +903,47 @@ static bool __attribute__((used))
...
+	// We add one element for the end-of-array indicator.
+	struct sc_mount *mounts = calloc(num_homedirs + 1, sizeof(struct sc_mount));
+	if (mounts == NULL) {
+		die("cannot allocate mount data for homedirs");
+	}
+
+    // Copy inv->homedirs to the mount structures
+	for (int i = 0; i < num_homedirs; i++) {
+		debug("Adding homedir: %s", inv->homedirs[i]);
+		mounts[i].path = sc_strdup(inv->homedirs[i]);
```
Comment (alexmurray) [inline cmd/snap-confine/mount-support.c:926]: Should `is_bidirectional` also be set here? I assume changes should be propagated back to the real home.

File: cmd/snap-confine/snap-confine-invocation.h [hunk 35]
```diff
@@ -38,6 +38,7 @@ typedef struct sc_invocation {
     /* Things derived at runtime. */
     char *base_snap_name;
     char *rootfs_dir;
+    char **homedirs;
```
Comment (alexmurray) [inline cmd/snap-confine/snap-confine-invocation.h:41]: Can you also add `num_homedirs` or similar which would then allow to remove the code which counts `homedirs` in `homedirs_are_mounted()` and `sc_homedir_mounts()` etc?

File: cmd/snap-confine/snap-confine-invocation.c [hunk 36]
```diff
@@ -144,3 +145,62 @@ void sc_check_rootfs_dir(sc_invocation *inv) {
...
+    size_t line_size = 0;
+    while (getline(&line, &line_size, f) != -1) {
+        if (sc_startswith(line, "homedirs=")) {
+            return sc_strdup(line + (sizeof("homedirs=") - 1));
+        }
+    }
+    return NULL;
+}
+
+void sc_invocation_check_homedirs(sc_invocation *inv)
```
Comment (alexmurray) [inline cmd/snap-confine/snap-confine-invocation.c:167]: since this modifies `inv` by setting up `homedirs` should it instead be called something like `sc_invocation_populate_homedirs`?

**Review by alexmurray — APPROVED** [review]
LGTM in general, but a few minor changes would make it a bit more readable IMO.

File: cmd/snap-confine/mount-support.c (hunk 34 above)
Comment (Meulengracht) [reply cmd/snap-confine/mount-support.c:921]: Really good point. Code updated.

File: cmd/snap-confine/snap-confine-invocation.h (hunk 35 above)
Comment (Meulengracht) [reply cmd/snap-confine/snap-confine-invocation.h:41]: Nice suggestion, it simplified the code a few places. Done

File: cmd/snap-confine/snap-confine-invocation.c (hunk 36 above)
Comment (Meulengracht) [reply cmd/snap-confine/snap-confine-invocation.c]: Yea that makes much more sense,  I renamed it to `sc_invocation_init_homedirs` to be in line with other namings
